File: 09-Recursion/java/01-SubsetSumEqualsToTarget.java
class Solution {
    public boolean isSubsetSum(int[] arr, int target) {
        int n = arr.length;
        boolean[][] dp = new boolean[n + 1][target + 1];

        for (int i = 0; i <= n; i++) {
            dp[i][0] = true;
        }

        for (int i = 1; i <= n; i++) {
            for (int sum = 1; sum <= target; sum++) {
                boolean notTake = dp[i - 1][sum];
                boolean take = false;

                if (arr[i - 1] <= sum) {
                    take = dp[i - 1][sum - arr[i - 1]];
                }

                dp[i][sum] = take || notTake;
            }
        }

        return dp[n][target];
    }
}

File: 09-Recursion/python/01_subset_sum_equals_to_target.py
class Solution:
    def isSubsetSum(self, arr: List[int], target: int) -> bool:
        n = len(arr)
        dp = [[False] * (target + 1) for _ in range(n + 1)]

        for i in range(n + 1):
            dp[i][0] = True

        for i in range(1, n + 1):
            for s in range(1, target + 1):
                not_take = dp[i - 1][s]
                take = False
                if arr[i - 1] <= s:
                    take = dp[i - 1][s - arr[i - 1]]
                dp[i][s] = take or not_take

        return dp[n][target]
